File: app/src/main/java/com/weatherkremenchug/weatherkremenchugtest/MainActivity.java
package com.weatherkremenchug.weatherkremenchugtest;

import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.SharedPreferences;
import android.content.pm.PackageManager;
import android.os.Build;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.design.widget.NavigationView;
import android.support.v4.app.ActivityCompat;
import android.support.v4.content.ContextCompat;
import android.support.v4.view.GravityCompat;
import android.support.v4.widget.DrawerLayout;
import android.support.v7.app.ActionBarDrawerToggle;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;
import android.view.MenuItem;
import android.view.View;
import android.widget.TextView;
import android.widget.Toast;

import com.weatherkremenchug.weatherkremenchugtest.activity.SettingsActivity;
import com.weatherkremenchug.weatherkremenchugtest.activity.WeatherOnTodayActivity;
import com.weatherkremenchug.weatherkremenchugtest.adapter.WeatherMainAdapter;
import com.weatherkremenchug.weatherkremenchugtest.data.ListData;
import com.weatherkremenchug.weatherkremenchugtest.data.WeatherForecast;
import com.weatherkremenchug.weatherkremenchugtest.retrofit.CallToServer;

import java.util.List;
import java.util.Objects;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class MainActivity extends AppCompatActivity
        implements NavigationView.OnNavigationItemSelectedListener {

    @SuppressWarnings("SpellCheckingInspection")
    private static final String APP_ID = "c3eebf803f44713f50e31a7c5b215a73";

    GPSTracker gpsTracker;

    double latitude;
    double longitude;

    String units = "metric";

    RecyclerView recyclerView;
    WeatherMainAdapter weatherMainAdapter;
    LinearLayoutManager linearLayoutManager;

    TextView log_out;

    SharedPreferences sharedPreferences;
    public static final String APP_PREFERENCES = "settings";

    AlertDialog.Builder alert;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT) {
            Objects.requireNonNull(getSupportActionBar()).setTitle("Погода на неделю");
        }

        try {
            if (ContextCompat.checkSelfPermission(getApplicationContext(),
                    android.Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED ) {
                ActivityCompat.requestPermissions(this,
                        new String[]{android.Manifest.permission.ACCESS_FINE_LOCATION}, 101);
            }
        } catch (Exception e){
            e.printStackTrace();
        }

        gpsTracker = new GPSTracker(MainActivity.this);

        if (gpsTracker.canGetLocation) {
            latitude = gpsTracker.getLatitude();
            longitude = gpsTracker.getLongitude();
        } else {
            gpsTracker.showSettingsAlert();
        }
        
        callWeather();

        sharedPreferences = getSharedPreferences(APP_PREFERENCES, Context.MODE_PRIVATE);
        final boolean switch_ask_exit_state = sharedPreferences.getBoolean("switch_ask_when_exit", false);

        recyclerView = findViewById(R.id.recycler_main);
        linearLayoutManager = new LinearLayoutManager(this);
        recyclerView.setLayoutManager(linearLayoutManager);

        log_out = findViewById(R.id.log_out);
        log_out.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                String title = getResources().getString(R.string.ask_exit);
                String button1 = getResources().getString(R.string.ask_exit_yes);
                String button2 = getResources().getString(R.string.ask_exit_no);

                if (switch_ask_exit_state) {
                    alert = new AlertDialog.Builder(MainActivity.this);
                    alert.setTitle(title);

                    alert.setPositiveButton(button1, new DialogInterface.OnClickListener() {
                        @Override
                        public void onClick(DialogInterface dialogInterface, int i) {
                            finish();
                        }
                    });

                    alert.setNegativeButton(button2, new DialogInterface.OnClickListener() {
                        @Override
                        public void onClick(DialogInterface dialogInterface, int i) {
                            alert.setCancelable(true);
                        }
                    });

                    alert.show();

                } else {
                    finish();
                }


            }
        });

       // Toast.makeText(getApplicationContext(), "Your Location is - \nLat: " + latitude + "\nLong: " + longitude, Toast.LENGTH_LONG).show();

        DrawerLayout drawer = findViewById(R.id.drawer_layout);
        ActionBarDrawerToggle toggle = new ActionBarDrawerToggle(
                this, drawer, toolbar, R.string.navigation_drawer_open, R.string.navigation_drawer_close);
        drawer.addDrawerListener(toggle);
        toggle.syncState();

        NavigationView navigationView = findViewById(R.id.nav_view);
        navigationView.setNavigationItemSelectedListener(this);
    }

    private void callWeather() {

        Retrofit retrofit = new Retrofit.Builder()
                .baseUrl("http://api.openweathermap.org/data/2.5/")
                .addConverterFactory(GsonConverterFactory.create())
                .build();

        CallToServer callToServer = retrofit.create(CallToServer.class);

        Call<WeatherForecast> call = callToServer.getForecast(latitude, longitude, units, APP_ID);
        call.enqueue(new Callback<WeatherForecast>() {
            @Override
            public void onResponse(@NonNull Call<WeatherForecast> call, @NonNull Response<WeatherForecast> response) {

                if (response.isSuccessful()) {
                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT) {
                        generateDataList(Objects.requireNonNull(response.body()).getList());
                    }
                }

                Toast.makeText(getApplicationContext(), "OK", Toast.LENGTH_LONG).show();
            }

            @Override
            public void onFailure(@NonNull Call<WeatherForecast> call, @NonNull Throwable t) {
                Toast.makeText(getApplicationContext(), "NOT_OK", Toast.LENGTH_LONG).show();
            }
        });

    }


    public void generateDataList(List<ListData> body) {
        weatherMainAdapter = new WeatherMainAdapter(this, body);
        recyclerView.setAdapter(weatherMainAdapter);
    }

    @Override
    public void onBackPressed() {
        DrawerLayout drawer = findViewById(R.id.drawer_layout);
        if (drawer.isDrawerOpen(GravityCompat.START)) {
            drawer.closeDrawer(GravityCompat.START);
        } else {
            super.onBackPressed();
        }
    }

//    @Override
//    public boolean onCreateOptionsMenu(Menu menu) {
//        // Inflate the menu; this adds items to the action bar if it is present.
//        getMenuInflater().inflate(R.menu.main, menu);
//        return true;
//    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        // Handle action bar item clicks here. The action bar will
        // automatically handle clicks on the Home/Up button, so long
        // as you specify a parent activity in AndroidManifest.xml.
        int id = item.getItemId();

        //noinspection SimplifiableIfStatement
        if (id == R.id.action_settings) {
            return true;
        }

        return super.onOptionsItemSelected(item);
    }

    @SuppressWarnings("StatementWithEmptyBody")
    @Override
    public boolean onNavigationItemSelected(@NonNull MenuItem item) {
        // Handle navigation view item clicks here.
        int id = item.getItemId();

        if (id == R.id.nav_weather_today) {
            Intent intent_weather_on_today = new Intent(MainActivity.this, WeatherOnTodayActivity.class);
            startActivity(intent_weather_on_today);
        } else if (id == R.id.nav_weather_week) {

        } else if (id == R.id.nav_drawer_settings) {
            Intent intent_settings = new Intent(MainActivity.this, SettingsActivity.class);
            startActivity(intent_settings);
            finish();
        }

        DrawerLayout drawer = findViewById(R.id.drawer_layout);
        drawer.closeDrawer(GravityCompat.START);
        return true;
    }

}

File: app/src/main/java/com/weather/weathermain/activity/main/MainActivity.kt
package com.weather.weathermain.activity.main

import android.arch.lifecycle.Observer
import android.arch.lifecycle.ViewModelProviders
import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import android.content.pm.PackageManager
import android.os.Build
import android.os.Bundle
import android.support.design.widget.NavigationView
import android.support.v4.app.ActivityCompat
import android.support.v4.content.ContextCompat
import android.support.v4.view.GravityCompat
import android.support.v4.widget.DrawerLayout
import android.support.v7.app.ActionBar
import android.support.v7.app.ActionBarDrawerToggle
import android.support.v7.app.AlertDialog
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.LinearLayoutManager
import android.util.Log
import android.view.Menu
import android.view.MenuItem
import com.weather.weathermain.GPSTracker
import com.weather.weathermain.R
import com.weather.weathermain.activity.main.adapter.WeatherMainAdapter
import com.weather.weathermain.activity.main.mainviewmodel.MainActivityViewModel
import com.weather.weathermain.activity.settings.SettingsActivity
import com.weather.weathermain.activity.weathertoday.CurrentWeatherActivity
import com.weather.weathermain.data.ListData
import com.weather.weathermain.data.WeatherForecast
import com.weather.weathermain.utils.constants.APP_ID
import com.weather.weathermain.utils.constants.APP_PREFERENCES
import com.weather.weathermain.utils.constants.UNITS
import kotlinx.android.synthetic.main.activity_main.*
import kotlinx.android.synthetic.main.app_bar_main.*
import kotlinx.android.synthetic.main.content_main.*
import java.util.*


class MainActivity : AppCompatActivity(), NavigationView.OnNavigationItemSelectedListener {
    companion object {
        fun launch(context: Context) {
            val launcher = Intent(context, MainActivity::class.java)
            context.startActivity(launcher)
        }
    }

    private lateinit var model: MainActivityViewModel

    private var gpsTracker: GPSTracker? = null

    private var latitude: Double = 0.toDouble()
    private var longitude: Double = 0.toDouble()

    private var weatherMainAdapter: WeatherMainAdapter? = null
    private var linearLayoutManager: LinearLayoutManager? = null

    private var sharedPreferences: SharedPreferences? = null

    private var alert: AlertDialog.Builder? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        setSupportActionBar(toolbar)
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT) {
            Objects.requireNonNull<ActionBar>(supportActionBar).title = "Погода на неделю"
        }

        try {
            if (ContextCompat.checkSelfPermission(applicationContext,
                            android.Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
                ActivityCompat.requestPermissions(this,
                        arrayOf(android.Manifest.permission.ACCESS_FINE_LOCATION), 101)
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }

        gpsTracker = GPSTracker(this@MainActivity)

        if (gpsTracker!!.canGetLocation) {
            latitude = gpsTracker!!.getLatitude()
            longitude = gpsTracker!!.getLongitude()
        } else {
            gpsTracker!!.showSettingsAlert()
        }

        model = ViewModelProviders.of(this)[MainActivityViewModel::class.java]
        model.requestForecastWeek(latitude, longitude, UNITS, APP_ID)

        model._getDataFail().observe(this, Observer<String> { data_fail ->
            Log.d("DATA_IS_FAIL ", data_fail.toString())
        })

        model._getForecastWeek().observe(this, Observer<WeatherForecast>{ data_forecast ->
            generateDataList(data_forecast!!.list)
        })

        sharedPreferences = getSharedPreferences(APP_PREFERENCES, Context.MODE_PRIVATE)
        val switchAskExitState = sharedPreferences!!.getBoolean("switchAskWhenExit", false)

        linearLayoutManager = LinearLayoutManager(this)
        recycler_main.layoutManager = linearLayoutManager

        log_out.setOnClickListener {
            val title = resources.getString(R.string.ask_exit)
            val button1 = resources.getString(R.string.ask_exit_yes)
            val button2 = resources.getString(R.string.ask_exit_no)

            if (switchAskExitState) {
                alert = AlertDialog.Builder(this@MainActivity)
                alert!!.setTitle(title)
                alert!!.setPositiveButton(button1) { dialogInterface, i -> finish() }
                alert!!.setNegativeButton(button2) { dialogInterface, i -> alert!!.setCancelable(true) }
                alert!!.show()

            } else {
                finish()
            }
        }

        val drawer = findViewById<DrawerLayout>(R.id.drawer_layout)
        val toggle = ActionBarDrawerToggle(
                this, drawer, toolbar, R.string.navigation_drawer_open, R.string.navigation_drawer_close)
        drawer.addDrawerListener(toggle)
        toggle.syncState()

        val navigationView = findViewById<NavigationView>(R.id.nav_view)
        navigationView.setNavigationItemSelectedListener(this)
    }

    fun generateDataList(body: List<ListData>) {
        weatherMainAdapter = WeatherMainAdapter(body)
        recycler_main!!.adapter = weatherMainAdapter
    }

    override fun onBackPressed() {
        val drawer = findViewById<DrawerLayout>(R.id.drawer_layout)
        if (drawer.isDrawerOpen(GravityCompat.START)) {
            drawer.closeDrawer(GravityCompat.START)
        } else {
            super.onBackPressed()
        }
    }

    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
        menuInflater.inflate(R.menu.main, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        val id = item.itemId
        return if (id == R.id.action_settings) {
            true
        } else super.onOptionsItemSelected(item)
    }

    override fun onNavigationItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.nav_weather_today -> CurrentWeatherActivity.launch(this)
            R.id.nav_weather_week -> { }
            R.id.nav_drawer_settings -> {
                SettingsActivity.launch(this)
                finish()
            }
        }

        val drawer = findViewById<DrawerLayout>(R.id.drawer_layout)
        drawer.closeDrawer(GravityCompat.START)
        return true
    }

}
//221